### PR TITLE
added gadget smaller

### PR DESCRIPTION
**Pull request by Barichek**

### CODE REVIEW

File: src/circuit/num.rs [hunk 1]
```diff
@@ -696,6 +696,48 @@ impl<E: Engine> AllocatedNum<E> {
         Ok(())
     }
 
+    /// Takes two allocated numbers (a, b) and returns
+    /// allocated boolean variable with value `true`
+    /// if the `a` smaller than `b`, `false` otherwise.
+    pub fn smaller<CS>(
```
Comment (gluk64) [inline src/circuit/num.rs]: smaller => less_than (commonly abbreviated as lt)

File: src/circuit/num.rs [hunk 2]
```diff
@@ -696,6 +696,48 @@ impl<E: Engine> AllocatedNum<E> {
...
+            || "delta equality",
+            |lc| lc + delta.get_variable(),
+            |lc| lc + CS::one(),
+            |lc| lc + a.get_variable() - b.get_variable()
+        );
+
+        /// Let 'bits' - 255 bits representation of delta
+        let bits = delta.into_bits_le(&mut cs)?;
+
+        match bits[254].get_variable() {
```
Comment (alexander-movchan) [inline src/circuit/num.rs]: 254 is not universal. Next test panics:

```rust
    #[test]
    fn test_less_than() -> Result<(), SynthesisError> {
        use bellman::pairing::bn256::{Bn256, Fr};

        let mut cs = TestConstraintSystem::<Bn256>::new();

        let a = AllocatedNum::alloc(cs.namespace(|| "a"), || Ok(Fr::from_str("10").unwrap()))?;
        let b = AllocatedNum::alloc(cs.namespace(|| "b"), || Ok(Fr::from_str("12").unwrap()))?;

        let num = AllocatedNum::smaller(cs.namespace(|| "a < b"), &a, &b)?;
        assert_eq!(num.get_value(), Some(true));

        Ok(())
    }
```

**Comment by alexander-movchan** [comment]
Doesn't work well when a number is as long as `Fr::CAPACITY`.
However, it may be used if we consider the numbers to be signed.

```rust
    #[test]
    fn test_less_than() -> Result<(), SynthesisError> {
        let mut cs = TestConstraintSystem::<Bls12>::new();

        let small = AllocatedNum::alloc(
            cs.namespace(|| "small"),
            || Ok(Fr::one()))?;

        let big = AllocatedNum::alloc(
            cs.namespace(|| "big"),
            || Ok(Fr::from_hex("73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000000").unwrap()))?;

        let should_be_true = AllocatedNum::smaller(cs.namespace(|| "small < big"), &small, &big)?;

        assert!(cs.is_satisfied());

        assert_eq!(should_be_true.get_value(), Some(true));

        Ok(())
    }
```

File: src/circuit/num.rs (hunk 1 above)
Comment (shamatar) [reply src/circuit/num.rs]: Also such functions is usually called “less than it equal” or “less” depending if it’s >= or >

File: src/circuit/num.rs (hunk 2 above)
Comment (shamatar) [reply src/circuit/num.rs]: Such check is proper for uintX arithmetic (where X is strictly less than bit length of the field), but I’m not sure about direct field elements. E.g for a prime 13: 0 - 12 = 1 and will have zero in the top bit

**Comment by shamatar** [comment]
Most likely it would require >2*N constraints where N is a bit length. Either bit decompose both numbers and just compare bit by bit, or bit decompose, then recompose N-1 highest bits, run the procedure for them using zeroness of the top bit of the difference (now it’s valid), and then also check the lowest bit. For recomposes numbers made from the highest bits you would have to keep both flags “less” and “equal” separately to accurately resolve the lowest bit

**Comment by Barichek** [comment]
used E::Fr::NUM_BITS in less_than gadget

This gadget takes E::Fr::NUM_BITS+2 constraints.

File: src/circuit/num.rs [hunk 3]
```diff
@@ -729,10 +729,12 @@ impl<E: Engine> AllocatedNum<E> {
...
         );
 
-        /// Let 'bits' - 255 bits representation of delta
+        /// Let 'bits' - E::Fr::NUM_BITS bits representation of delta
         let bits = delta.into_bits_le(&mut cs)?;
 
-        match bits[254].get_variable() {
+        assert_eq!(bits.len() as u32, E::Fr::NUM_BITS);
+
+        match bits[(E::Fr::NUM_BITS as usize) - 1].get_variable() {
```
Comment (shamatar) [inline src/circuit/num.rs]: How does this Not crash? But decomposition should give a vector of NUM_BITS elements which is zero enumerates, so you actually try to seek pass the end

**Pull request closed without merging**